From c98b641fc43716d325f04a68c4a181a04d5838a3 Mon Sep 17 00:00:00 2001
From: shulbao <quic_shulbao@quicinc.com>
Date: Mon, 20 Jan 2025 14:25:42 +0800
Subject: [PATCH] monitor soundcard to recover init state

[PATCH] monitor soundcard to recover init state after SSR
monitor sound card state to revocer pulseaudio init state
after SSR, to fix issue that pluseaudio can't run usecase after ssr.

Upstream-Status: Inappropriate [oe]
Signed-off-by: shulbao <quic_shulbao@quicinc.com>
---
 src/pulse/mainloop.c | 64 ++++++++++++++++++++++++++++++++++++++++++++
 src/pulse/mainloop.h |  8 +++++-
 2 files changed, 71 insertions(+), 1 deletion(-)

diff --git a/src/pulse/mainloop.c b/src/pulse/mainloop.c
index 90112f5..10a0255 100644
--- a/src/pulse/mainloop.c
+++ b/src/pulse/mainloop.c
@@ -53,6 +53,14 @@
 #include "mainloop.h"
 #include "internal.h"
 
+#include <sys/stat.h>
+#include <pthread.h>
+#include <ctype.h>
+#define CARD_STATUS_NONE   -1
+#define CARD_STATUS_OFFLINE 0
+#define CARD_STATUS_ONLINE  1
+pthread_t thread;
+
 struct pa_io_event {
     pa_mainloop *mainloop;
     bool dead:1;
@@ -130,8 +138,60 @@ struct pa_mainloop {
     pa_poll_func poll_func;
     void *poll_func_userdata;
     int poll_func_ret;
+    FILE* fp;
 };
 
+FILE* open_soundcard() {
+    FILE *fp;
+    fp = open("/sys/kernel/snd_card/card_state", O_RDONLY);
+	if (fp < 0){
+        pa_log("Failed to open snd card ");
+		return NULL;
+	}
+    return fp;
+}
+
+int check_soundcard(FILE* fp) {
+    char *sndcard_state = "";
+	int status = CARD_STATUS_NONE;
+	int pre_status = CARD_STATUS_ONLINE;
+	struct stat buf;
+	void* hook_data;
+	void* slot_data;
+	void* call_data;
+	char scan_buf[10];
+	int card_status = 0;
+
+    if(!fp) {
+        pa_log("Failed to check sound card state" );
+        return 2;
+    }
+    memset(scan_buf ,0 ,sizeof(scan_buf));
+    read(fp, scan_buf, 10);
+    lseek(fp, 0L, SEEK_SET);
+    sscanf(scan_buf, "%d", &card_status);
+    if (card_status == 0) {
+		status = CARD_STATUS_OFFLINE;
+    } else if (card_status == 1) {
+        status = CARD_STATUS_ONLINE;
+    } else {
+        pa_log_info("unknown state, %d", status);
+    }
+    if (status == pre_status) {
+        /* nothing change,,no action*/
+    } else if (status == CARD_STATUS_OFFLINE) {
+        //ssr down. need stop current stream
+        pa_log_info("sound card status change to offline");
+        pre_status = status;
+        return -1;
+    } else if (status == CARD_STATUS_ONLINE) {
+        pa_log_info("sound card status change to online");
+        pre_status = status;
+        return 1;
+    }
+    usleep(1000);
+    return 1;
+}
 static short map_flags_to_libc(pa_io_event_flags_t flags) {
     return (short)
         ((flags & PA_IO_EVENT_INPUT ? POLLIN : 0) |
@@ -948,6 +1008,8 @@ int pa_mainloop_iterate(pa_mainloop *m, int block, int *retval) {
     if ((r = pa_mainloop_dispatch(m)) < 0)
         goto quit;
 
+    r = check_soundcard(m->fp);
+
     return r;
 
 quit:
@@ -960,6 +1022,8 @@ quit:
 int pa_mainloop_run(pa_mainloop *m, int *retval) {
     int r;
 
+    m->fp = open_soundcard();
+
     while ((r = pa_mainloop_iterate(m, 1, retval)) >= 0)
         ;
 
diff --git a/src/pulse/mainloop.h b/src/pulse/mainloop.h
index f1069da..d3fc848 100644
--- a/src/pulse/mainloop.h
+++ b/src/pulse/mainloop.h
@@ -20,7 +20,7 @@
   You should have received a copy of the GNU Lesser General Public License
   along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
 ***/
-
+#include <stdio.h>
 #include <pulse/mainloop-api.h>
 #include <pulse/cdecl.h>
 
@@ -129,6 +129,12 @@ typedef int (*pa_poll_func)(struct pollfd *ufds, unsigned long nfds, int timeout
 /** Change the poll() implementation */
 void pa_mainloop_set_poll_func(pa_mainloop *m, pa_poll_func poll_func, void *userdata);
 
+/** open soundcard */
+FILE* open_soundcard();
+
+/** check state of sound card */
+int check_soundcard(FILE* fp);
+
 PA_C_DECL_END
 
 #endif
-- 
2.25.1

